### PR TITLE
Modify monitor to allow arbitrary tests

### PR DESCRIPTION
**Pull request by benbarclay**
This change does a few things:

- Allows us to pass in an arbitrary ServiceDescriptor as a string
- No longer deep checks the generated ServiceInstance(s) as we don't know what will be generated
- Moves the checks to using polling everywhere and uses the ExpectedProcessingTime flag as a deadline for both the test + cleanup

### CODE REVIEW

File: pkg/monitor/monitor.go [hunk 1]
```diff
@@ -228,7 +188,7 @@ func (m *Monitor) logFailureDetails(ctx context.Context, retErr error) {
 		service = Service(srv)
 	}
 
-	m.Logger.Error("monitor job failure", serviceInstance, serviceDescriptor, service, zap.Error(retErr))
+	m.Logger.Error("monitor job failure", serviceDescriptor, service, zap.Error(retErr))
```
Comment (nilebox) [inline pkg/monitor/monitor.go]: Shall we log summary from reporter as well? :philosoraptor:

File: cmd/monitor/app/app.go [hunk 2]
```diff
@@ -18,9 +18,10 @@ import (
 )
 
 type App struct {
-	RestConfig *rest.Config
-	Logger     *zap.Logger
-	Options    Options
+	RestConfig        *rest.Config
+	Logger            *zap.Logger
+	Options           Options
+	ServiceDescriptor string
```
Comment (kopper) [inline cmd/monitor/app/app.go]: nit: it would be nicer if this is part of Options and unmarshalled and parsed to from flag

File: pkg/monitor/monitor.go [hunk 3]
```diff
@@ -204,14 +172,6 @@ func (m *Monitor) waitForServiceDescriptorDeletion(ctx context.Context, name str
 }
 
 func (m *Monitor) logFailureDetails(ctx context.Context, retErr error) {
-	var serviceInstance zap.Field
```
Comment (kopper) [inline pkg/monitor/monitor.go]: we need to log these details

Comment (kopper) [reply pkg/monitor/monitor.go]: it should be good enough to log full SD and LD as LD contains detailed failure reasons for individual resources

Comment (benbarclay) [reply pkg/monitor/monitor.go]: Yeah, the problem is we don't know what ServiceInstances exist for a given ServiceDescriptor. By removing this code path we just always log the full SD and any errors associated with it.

Comment (kopper) [reply pkg/monitor/monitor.go]: 1. Of course we do know. If reporter is able to figure out SIs we are able to do it too.
2. We don't need to because all details are in LD anyways.

Comment (kopper) [reply pkg/monitor/monitor.go]: I mean there might be some obscure cases when Voyager is so broken than LD has fake news in it, but 🤷‍♂️

Comment (nilebox) [reply pkg/monitor/monitor.go]: @benbarclay as I pointed above, we could just log reporter output, but it's going to be HUGE. Adding LD is probably good enough actually, as it contains bubbled up errors from ServiceInstances etc (but SD doesn't!)

File: cmd/monitor/app/app.go (hunk 2 above)
Comment (ash2k) [reply cmd/monitor/app/app.go]: I second that. And `App` should not be parsing it from string, it should accept a typed object.

File: pkg/monitor/monitor.go [hunk 4]
```diff
@@ -75,67 +74,36 @@ func (m *Monitor) Run(ctx context.Context) (retErr error) {
...
-		return errors.Errorf("version parameter missing from ServiceInstance %q in %q", resourceName, namespace)
-	}
-	if foundVersion != expectedVersion {
-		return errors.Errorf("found ServiceInstance with embedded version %q but expecting %q", foundVersion, expectedVersion)
-	}
+		_, ready := cond_v1.FindCondition(sd.Status.Conditions, cond_v1.ConditionReady)
+		if ready == nil {
+			// Probably fine, this can trigger before the SD is processed by the cluster
+			return false, nil
+		}
```
Comment (ash2k) [inline pkg/monitor/monitor.go]: generation and observed generation should be checked here in the future.

Comment (nilebox) [reply pkg/monitor/monitor.go]: @ash2k SD status still doesn't have `ObservedGeneration` AFAICT :(

File: pkg/monitor/monitor.go (hunk 3 above)
Comment (benbarclay) [reply pkg/monitor/monitor.go]: I'll add LD for now, but that just seems like a hack around the problems with error propagation to SD

File: pkg/monitor/monitor.go (hunk 4 above)
Comment (ash2k) [reply pkg/monitor/monitor.go]: yes, it does not. That is why I'm saying "in the future" :)

File: pkg/monitor/monitor.go (hunk 3 above)
Comment (ash2k) [reply pkg/monitor/monitor.go]: It's not a hack, it's by design. See the relevant doc. You were on that meeting btw 😴

Comment (amckague) [reply pkg/monitor/monitor.go]: Dumping reporter output is preferable to reaching for other objects.

File: pkg/monitor/monitor.go [hunk 5]
```diff
@@ -228,7 +186,7 @@ func (m *Monitor) logFailureDetails(ctx context.Context, retErr error) {
 		service = Service(srv)
 	}
 
-	m.Logger.Error("monitor job failure", serviceInstance, serviceDescriptor, service, zap.Error(retErr))
+	m.Logger.Error("monitor job failure", serviceDescriptor, service, zap.Error(retErr))
 }
 
 func (m *Monitor) serviceInstance(ctx context.Context, name, namespace string) (*v1beta1.ServiceInstance, error) {
```
Comment (ychen-atlassian) [inline pkg/monitor/monitor.go]: Remove this method if it's no longer being used

File: pkg/monitor/monitor.go (hunk 3 above)
Comment (benbarclay) [reply pkg/monitor/monitor.go]: According to the spec, the SD status is supposed to have the status of each location. If we're dropping the correct status of a location such that we need to pull the LD to get that information, then it's an issue and this is a hack.

File: pkg/monitor/monitor.go (hunk 5 above)
Comment (benbarclay) [reply pkg/monitor/monitor.go]: done

File: pkg/monitor/monitor.go (hunk 3 above)
Comment (nilebox) [reply pkg/monitor/monitor.go]: I tend to agree with Ben. Even if at first we will only show a cluster-local status (i.e. it won't include status of LD from the secondary cluster), it will be a huge improvement already.

Comment (kopper) [reply pkg/monitor/monitor.go]: SD contains status of each location, but no resource statuses. Thus it is useless.

File: cmd/monitor/app/app.go [hunk 6]
```diff
@@ -18,9 +19,10 @@ import (
...
```
Comment (kopper) [inline cmd/monitor/app/app.go]: not used anymore?

Comment (benbarclay) [reply cmd/monitor/app/app.go]: Good catch

**Comment by kopper** [comment]
This PR when merged will break deployments? (because there is no config yet)

**Comment by benbarclay** [comment]
I plan to merge this into the upstream repo with the config files at the same time

File: cmd/monitor/app/options.go [hunk 7]
```diff
@@ -4,17 +4,19 @@ import (
...
 	"sigs.k8s.io/yaml"
 )
 
 type Options struct {
-	ServiceDescriptorName  string                 `json:"name"`
-	Location               options.Location       `json:"location"`
-	ExpectedProcessingTime time.Duration          `json:"expected_processing_time"`
-	ServiceSpec            creator_v1.ServiceSpec `json:"service_spec"`
+	ServiceDescriptorName  string                     `json:"name"`
+	ServiceDescriptor      *comp_v1.ServiceDescriptor `json:"service_descriptor"`
```
Comment (ash2k) [inline cmd/monitor/app/options.go]: Ideally we should use camel case everywhere but I see that is already not the case :( Would be great to change that in a separate PR.

File: pkg/monitor/monitor.go [hunk 8]
```diff
@@ -62,80 +67,43 @@ func (m *Monitor) Run(ctx context.Context) (retErr error) {
...
 	}
 
 	m.Logger.Info("Initialized ServiceDescriptor for synthetic checks")
 
-	// Wait until the corresponding ServiceInstance is created.
-	// Sleeps for a short period of time, which for now should be enough I think
-	time.Sleep(m.ExpectedProcessingTime)
-
-	err = m.verifyUpsServiceInstance(ctx, m.ServiceDescriptorName, m.Version)
+	err = wait.PollImmediate(pollDelay, m.ExpectedProcessingTime, m.verifyServiceDescriptorStatus(m.ServiceDescriptorName))
```
Comment (ash2k) [inline pkg/monitor/monitor.go]: We can avoid polling to get immediate results using `UntilWithSync()` from `"k8s.io/client-go/tools/watch"`. See how it is used in Smith's integration tests https://github.com/atlassian/smith/blob/e367e27c0d09db8c88715fd76122be086bacd15a/it/utils_for_tests.go#L89-L99
